File: 207-course-schedule/course-schedule.cpp
class Solution {
public:
    vector<int> BFS(int u, vector<int>&vis,vector<int>&indegree,unordered_map<int,vector<int>>&adj){
        queue<int>q;
        for(int i=0;i<indegree.size();i++) if(indegree[i]==0) q.push(i);
        vector<int>ans;
        while(!q.empty()){
            int curr = q.front();q.pop();
            ans.push_back(curr);
            for(int &v:adj[curr]){
                indegree[v]--;
                if(indegree[v]==0){
                    vis[v]=true;
                    q.push(v);
                }
            }
        }
        return ans;
    }
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int,vector<int>>adj;
        vector<int>indegree(numCourses,0);
        vector<int>vis(numCourses,0);
        for(vector<int>&prerequisite:prerequisites){
            adj[prerequisite[1]].push_back(prerequisite[0]);;
            indegree[prerequisite[0]]++;
        }

        vector<int> ans = BFS(0,vis,indegree,adj);
        return ans.size()==numCourses;
    }
};

File: 207-course-schedule/course-schedule.py
from collections import deque, defaultdict
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adj = defaultdict(list)
        indegree = [0]*numCourses
        
        for v,u in prerequisites:
            adj[u].append(v)
            indegree[v]+=1

        q = deque()

        for i in range(numCourses):
            if indegree[i]==0:
                q.append(i)

        ans = []        
        while q:
            curr_course = q.popleft()
            ans.append(curr_course)

            for next_course in adj[curr_course]:
                indegree[next_course]-=1
                if indegree[next_course] == 0:
                    q.append(next_course)

        return len(ans) == numCourses           


        